### ISSUE TITLE
[4.0.0.beta-17] ActivatedRoute.queryParams break after navigation

### ISSUE THREAD

**Issue by h4rm** [ionic-team/ionic#16736]
# Bug Report

**Ionic version:**
4.0.0.beta-17 and 4.0.0.beta-18

**Current behavior:**
Subscriptions on ActivatedRoute.queryParams do not fire after navigating away from and returning to that page. This is particullary problematic for base pages of tabs that are cached. The problem does not occur on sub-pages because they are re-initialized i.e. onInit fires again and again (e.g. on speaker-detail in ionic-conference app).

**Expected behavior:**
Navigation should not interfere with the queryParams subscription.

**Steps to reproduce:**
1. Clone ionic-conference-app
2. Add code below
3. Test with "Test" button on speakers list. Observe console log of queryParams. Navigate to speaker detail and back. Try "Test" button again and observe no console output.
4. Test same with Test button speaker detail page. Navigate from base to detail and back and to detail again - logs are fired.

**Related code:**

`speaker-list.html`
```...
<ion-content class="outer-content">
   <!-- TEST BUTTON -->
  <ion-button (click)="test()">Test</ion-button>

  <ion-list>
...
```

in `speaker-list.ts` modify
```
...
export class SpeakerListPage implements OnInit {
  speakers: any[] = [];

  constructor(
    public actionSheetCtrl: ActionSheetController,
    public confData: ConferenceData,
    public inAppBrowser: InAppBrowser,
    private route: ActivatedRoute,
    private router: Router
  ) {}

  ngOnInit() {
    console.log('Init');
    this.route.queryParams.subscribe(test => console.log(test));
  }

  test() {
    this.router.navigate(['./'], { queryParams: { test: Math.random() }, relativeTo: this.route });
  }
...
```

**Ionic info:** 
```
Ionic:

   ionic (Ionic CLI)             : 4.5.0 (C:\Users\Ferdinand\AppData\Roaming\npm\node_modules\ionic)
   Ionic Framework               : @ionic/angular 4.0.0-beta.17
   @angular-devkit/build-angular : 0.10.7
   @angular-devkit/schematics    : 7.1.3
   @angular/cli                  : 7.1.3
   @ionic/angular-toolkit        : 1.2.0

Cordova:

   cordova (Cordova CLI) : 8.1.2 (cordova-lib@8.1.1)
   Cordova Platforms     : android 7.1.4
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 2.2.5, (and 8 other plugins)

System:

   NodeJS : v10.11.0 (C:\Program Files\nodejs\node.exe)
   npm    : 6.4.1
   OS     : Windows 10
```

**Comment by h4rm** [comment]
Problem persists if `this.route.queryParams.subscribe(test => console.log(test));` is moved to `ionViewDidEnter`. The whole subscription seems to be corrupted.

**Comment by paulstelzer** [comment]
Thanks for your issue! Could you please create a repo? On my app (I also copied your snippet) this is working without problems.

**Comment by h4rm** [comment]
The problem is that the base page of the tab is cached (as it should) and when one navigates away and back to that page, the cache is retrieved but somehow the `queryParams` of that `ActivatedRoute` is not firing anymore, or maybe ActivatedRoute is not restored correctly.
I have the feeling that it might be related to #16534 where ActivatedRoute also introduces problems when the currect href is rendered. Although in that case, ActivatedRoute is sometimes not correctly restored when using relative navigation, which is not the case here.

**Comment by blackfan23** [comment]
@paulstelzer 
I provided a repo here with the changes described by @h4rm 

https://github.com/blackfan23/conference_app_beta_nav_bug

**Comment by paulstelzer** [comment]
The problem of the issue is the following:

Ionic puts the page into the stack, means the component will be not destroyed, instead it's just hidden (Read more here https://medium.com/@paulstelzer/ionic-4-and-the-lifecycle-hooks-4fe9eabb2864).

That's the reason why ngOnInit() will not be fired and I think it's the same why the router not working as you expected because on going back to the page, the page will be shown again (not created, it was already created - it will only be shown). And I think this leads that the queryParams not working as you expected

A workaround would be to avoid the ionic stack functionality and just navigate by root

**Comment by h4rm** [comment]
Hey Paul,

thanks for your remarks, What exactly do you mean by "navigate by root". If it meants that we avoid relative routes as used above, I have also tried to navigate via absolute roots
```
    this.router.navigate(['app', 'tabs', { outlets: { speakers: ['speakers'] } }], { queryParams: { test: Math.random() } });
```
and the problem persists.

Also, I understand that ionic only hides the base page and shows it again when navigating back. Unfortunately that does not explain why the first subscription in `ngOnInit()` does not fire again as the page component still exists in the background the the subscription should be still valid.

**Comment by paulstelzer** [comment]
As said you are the special case (like on my app) that you are stay on the same page. I presented my workaround in my comment above.

The  first subscription should work, I get the params on first load in my app.

**Comment by h4rm** [comment]
Okay, we found a viable work-around.

If we put the the queryParams subscription into a seperate service

```
import { Injectable } from '@angular/core';
import { ActivatedRoute } from '@angular/router';

@Injectable({
  providedIn: 'root',
})
export class TestService {
  constructor(private route: ActivatedRoute) {
    this.route.queryParams.subscribe(test => console.log(test));
  }
}
```

it will react to all param changes as expected. In particular, in the example above "when you use "test" params on the base page, navigate to a child page, put a link there back to the base page with another test=Math.Random()" - the queryParams correctly fires.
It does not fire, if the subscription is done in the ngOnInit which is the core problem that I do not understand.

**Comment by KevinKelchen** [comment]
Thanks for the quick response, @paulstelzer!

I don't _think_ I share the same issue as expressed in 16744. The issue I mention with the `ion-back-button` in 16516 (arguably there are two separate but very related issues mentioned in there) is that it doesn't work like the browser's history stack, Android back button, nor Ionic 3 when you use it to navigate back after having forward navigated to a cached Page in the stack. Forward navigating to a cached Page in the stack pops off all of the Pages in between. Not only do those Pages get destroyed, but the `ion-back-button` doesn't take you back to any of them--it takes you to back to the Page prior to the cached instance of the Page--wherever that was in the stack--so you don't go back to the Page you were just on prior to forward navigating to the cached Page. I'm not sure what the design rationale was behind this behavior or why it's not configurable to work like it used to. This is why I'm instead using a custom back button Component that simply calls Angular's `Location.back()`. This is not the place to really discuss my `ion-back-button` issues however--issue 16516 is. Just summarizing the distinction between issues here.

So for me, the new query parameter value _is_ present in the URL and it appears that the router is trying to set the observable values but the subscriptions aren't firing.

**Comment by daem0ndev** [comment]
I can confirm this issue as this was quite a blocker for us, the problem is that the OutletInjector class inside ion-router-outlet only provides the correct ActivatedRoute during initial component initialization, but this is not the correct ActivatedRoute when re-activating a component thats already in the ion-router-outlet stack. The fact of the matter is, there is no easy way to correct this issue from Ionic's side because you cant change the injected activated route after the fact. In other words, the activated route thats injected on a component is only correct if the component is being created by the factory, NOT if we are reactivating a component thats already created, and navigating BACK to that component is actually a whole new navigation and therefor new ActivatedRoute object.

I implemented a very crude workaround to this issue because we have a tight deadline to release our app. The main gist of the solution is that I created a service that always maintains the current activated route as well as the component that is activated and exposes this via an observable, and then on my page components I use this service to get the latest (and correct) instance of ActivatedRoute as it changes over time (i.e. when navigating BACK to a component on the stack) and then you can correctly switchMap over to the queryParamMap on that new ActivatedRoute.

**Comment by biesbjerg** [comment]
@daem0ndev can you share your workaround service?

**Comment by daem0ndev** [comment]
@biesbjerg for sure, do you have a repo with this issue that I can simply push a branch to?

**Comment by KevinKelchen** [comment]
Are there any thoughts from the team on when this might be fixed? Prior to Final? Seems pretty important for using the correct navigation parameters if the recommendation and default is to use cached Pages.

**Comment by mhartington** [comment]
Hey everyone. We just merged in #17888 which fixes this. Thanks for all your patience here 😄
